compiler/rustc_hir_typeck/src/expr_use_visitor.rs

```diff
@@ -80,21 +86,16 @@ pub trait Delegate<'tcx> {
     );
 }
 
-#[derive(Copy, Clone, PartialEq, Debug)]
-enum ConsumeMode {
-    /// reference to x where x has a type that copies
-    Copy,
-    /// reference to x where x has a type that moves
-    Move,
-}
-
 /// The ExprUseVisitor type
 ///
 /// This is the code that actually walks the tree.
-pub struct ExprUseVisitor<'a, 'tcx> {
-    mc: mc::MemCategorizationContext<'a, 'tcx>,
+pub struct ExprUseVisitor<'a, 'tcx, D: Delegate<'tcx>> {
+    typeck_results: &'a ty::TypeckResults<'tcx>,
+    infcx: &'a InferCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    upvars: Option<&'tcx FxIndexMap<HirId, hir::Upvar>>,
     body_owner: LocalDefId,
-    delegate: &'a mut dyn Delegate<'tcx>,
+    delegate: RefCell<&'a mut D>,
```

Comment:
Sorry this commit is so complicated. tl;dr is:
* All the methods from `MemCategorizationContext` got moved up into `ExprUseVisitor`
* Everything is `&self` now
* Delegate is a `RefCell`, and doesn't use dyn dispatch anymore
* `McResult` returns `ErrorGuaranteed` as previously asked